Fix TestStoreRangeRebalance

mrtracy opened the pull request:
Fixes #3483
Fixes #3092

Two commits: 
1. fixes an issue which arises in multiTestContext when the stores[0] is _not_ the range leader for the first range.
2. Increase a timeout in replicateRange() which was causing 1/5000 failures on local stress test runs.

Unskips TestStoreRangeRebalance.

<!-- Reviewable:start -->

[<img src="https://reviewable.io/review_button.png" height=40 alt="Review on Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/3814)

<!-- Reviewable:end -->

Commit:
multiTestContext now implements RangeDescriptorDB

Fixes #3483

`multiTestContext`, when initializing its DistSender, set the RangeDescriptorDB
for that sender to always use the first store in the context, `mtc.stores[0]`.

This was chosen because, unlike a running Server, the DistSender could not use
itself as a RangeDescriptorDB because the gossip network used in a
multiTestContext is only partially operational.

This functioned the majority of the time, but would break if the first range was
unreplicated from `stores[0]` OR if the replica of the first range on
`stores[0]` was no longer the range leader. This could cause tests to time out
if either situation was encountered.

To remedy this, multiTestContext now implements RangeDescriptorDB. DistSender's
implementation of RangeLookup() works correctly, but now there is a custom
implementation of FirstRange() which can reliably find the current
RangeDescriptor for the first range without needing the gossip network.

Commit also modifies multiTestContext.unreplicateRange() to wait until the
replica descriptor has been updated on the source node.

Commit:
Bump timeout, enable TestStoreRangeRebalance

Fixes #3092

This commit bumps the replicateRange timeout in `client_test.go`, which allows
the `replicateRange` function to take slightly longer to successfully replicate
a range. This appears to remove all random failures from local machine
high-stress test runs.

Commit also re-enables disabled test TestStoreRangeRebalance, which was disabled
due to this timeout and issue #3483

tamird commented:
Reviewed 2 of 2 files at r1, 2 of 2 files at r2.
Review status: all files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[storage/client_raft_test.go, line 596 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3814#-K7x16mgDe8oWC8oLJB9:-K7x16mgDe8oWC8oLJBA:-1079555497) ([raw file](https://github.com/cockroachdb/cockroach/blob/dc026f25e79da55f61473d8361bfce65e3acd483/storage/client_raft_test.go#L596)):*
teh pwnz0rz

---

*[storage/client_raft_test.go, line 612 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3814#-K7x1XxRPxcUrc9Xb029:-K7x1XxRPxcUrc9Xb02A:-214955580) ([raw file](https://github.com/cockroachdb/cockroach/blob/dc026f25e79da55f61473d8361bfce65e3acd483/storage/client_raft_test.go#L612)):*
I can't interpret this comment. It will fail, and fail the test? or do you mean it will fail but not block indefinitely, and thus do what you expect?

---

*[storage/client_test.go, line 377 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3814#-K7x1xYS1Y-H2mN_ayBb:-K7x1xYS1Y-H2mN_ayBc:524165196) ([raw file](https://github.com/cockroachdb/cockroach/blob/dc026f25e79da55f61473d8361bfce65e3acd483/storage/client_test.go#L377)):*
s/descriptors/replicas/? also, is this comment inverted? it seems that the logic above is sorting oldest-first, while this is sorting least-replicas-first, which would mean that more replicas is newer.

---

*[storage/client_test.go, line 382 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3814#-K7x2Obouq6A1UARA1Ks:-K7x2Obouq6A1UARA1Kt:87307939) ([raw file](https://github.com/cockroachdb/cockroach/blob/dc026f25e79da55f61473d8361bfce65e3acd483/storage/client_test.go#L382)):*
can you make "KeyMin" a reference to a real constant? i think you meant roachpb.KeyMin.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3814)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
Review status: all files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---

*[storage/client_test.go, line 378 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3814#-K7x5F0cGYCFwIU_NjDX:-K7x5F0cGYCFwIU_NjDY:-650557775) ([raw file](https://github.com/cockroachdb/cockroach/blob/dc026f25e79da55f61473d8361bfce65e3acd483/storage/client_test.go#L378)):*
One of these `i`s needs to be a `j`. The first one, I think. It would be good (and easy) to test `rangeDescByAge` in isolation.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3814)*
<!-- Sent from Reviewable.io -->

mrtracy commented:
Review status: 0 of 2 files reviewed at latest revision, 5 unresolved discussions.

---

*[storage/client_raft_test.go, line 596 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3814#-K7x16mgDe8oWC8oLJB9:-K7xF6b0LPDYLKlpfwZB:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/dc026f25e79da55f61473d8361bfce65e3acd483/storage/client_raft_test.go#L596)):*
Done.

---

*[storage/client_raft_test.go, line 612 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3814#-K7x1XxRPxcUrc9Xb029:-K7xFSVfTHOBIm9cmedB:-875515351) ([raw file](https://github.com/cockroachdb/cockroach/blob/dc026f25e79da55f61473d8361bfce65e3acd483/storage/client_raft_test.go#L612)):*
The test was created by Bram in order to demonstrate the failure; I needed to update the comment now that it no longer fails. Updated!

---

*[storage/client_test.go, line 377 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3814#-K7x1xYS1Y-H2mN_ayBb:-K7xNdPFFntq4pTyC5Xl:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/dc026f25e79da55f61473d8361bfce65e3acd483/storage/client_test.go#L377)):*
Done.

---

*[storage/client_test.go, line 378 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3814#-K7x5F0cGYCFwIU_NjDX:-K7xNUlrNkeD9F9YJbi_:-1155640519) ([raw file](https://github.com/cockroachdb/cockroach/blob/dc026f25e79da55f61473d8361bfce65e3acd483/storage/client_test.go#L378)):*
Done, and added a test.

---

*[storage/client_test.go, line 382 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3814#-K7x2Obouq6A1UARA1Ks:-K7xNcQgKCSNY95XaXbF:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/dc026f25e79da55f61473d8361bfce65e3acd483/storage/client_test.go#L382)):*
Done.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3814)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
LGTM

---

Review status: 0 of 2 files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3814#-:-K7xQTL7Gi-1ggV0Luan:676834233)*
<!-- Sent from Reviewable.io -->

tamird commented:
LGTM

---

Reviewed 2 of 2 files at r3.
Review status: all files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[storage/client_test.go, line 59 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3814#-K7xQtY1nZhLz07wqV_p:-K7xQtY1nZhLz07wqV_q:1528136898) ([raw file](https://github.com/cockroachdb/cockroach/blob/8906cc056aeb001b7ca029b7436c015a83733866/storage/client_test.go#L59)):*
nit: move the non-cockroach imports to a separate group above

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3814#-:-K7xRVn3fmv0It7Q6Qz-:676834233)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
needs `go generate ./...`.

mrtracy commented:
Fixed the `go generate`

---

Review status: 0 of 2 files reviewed at latest revision, 2 unresolved discussions.

---

*[storage/client_test.go, line 59 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3814#-K7xQtY1nZhLz07wqV_p:-K80Nerrfc3WgU2T1Z7B:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/8906cc056aeb001b7ca029b7436c015a83733866/storage/client_test.go#L59)):*
Done.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3814#-:-K80NfZN0-jGbeDX48hG:1226028755)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 2 of 2 files at r4.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3814)*
<!-- Sent from Reviewable.io -->
